src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs

```diff
@@ -0,0 +1,82 @@
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract class AbstractRecommendationServiceBasedCompletionProvider : AbstractSymbolCompletionProvider
+    {
+        protected override Task<IEnumerable<ISymbol>> GetSymbolsWorker(AbstractSyntaxContext context, int position, OptionSet options, CancellationToken cancellationToken)
+        {
+            var recommender = context.GetLanguageService<IRecommendationService>();
+            return recommender.GetRecommendedSymbolsAtPositionAsync(context.Workspace, context.SemanticModel, position, options, cancellationToken);
+        }
+
+        protected override async Task<IEnumerable<ISymbol>> GetPreselectedSymbolsWorker(AbstractSyntaxContext context, int position, OptionSet options, CancellationToken cancellationToken)
+        {
+            var recommender = context.GetLanguageService<IRecommendationService>();
+            var typeInferrer = context.GetLanguageService<ITypeInferenceService>();
+
+            var inferredTypes = typeInferrer.InferTypes(context.SemanticModel, position, cancellationToken)
+                ?.Where(t => t.SpecialType != SpecialType.System_Void)
+                .ToSet();
+            if (inferredTypes == null || !inferredTypes.Any())
+            {
+                return SpecializedCollections.EmptyEnumerable<ISymbol>();
+            }
+
+            var symbols = await recommender.GetRecommendedSymbolsAtPositionAsync(context.Workspace, context.SemanticModel, position, options, cancellationToken).ConfigureAwait(false);
+            return symbols.Where(s => inferredTypes.Contains(s.GetSymbolType()) && !IsInstrinsic(s));
+        }
+
+        protected override CompletionItem CreateItem(string displayText, string insertionText, int position, List<ISymbol> symbols, AbstractSyntaxContext context, TextSpan span, bool preselect, SupportedPlatformData supportedPlatformData)
+        {
+            var matchPriority = preselect ? ComputeSymbolMatchPriority(symbols[0]) : MatchPriority.Default;
+
+            return SymbolCompletionItem.Create(
+                            displayText: displayText,
+                            insertionText: insertionText,
+                            filterText: GetFilterText(symbols[0], displayText, context),
+                            span: span,
+                            contextPosition: context.Position,
+                            descriptionPosition: position,
+                            symbols: symbols,
+                            supportedPlatforms: supportedPlatformData,
+                            matchPriority: matchPriority,
+                            rules: GetCompletionItemRules(symbols, context));
+        }
+
+        protected abstract bool IsInstrinsic(ISymbol s);
+
+        private static MatchPriority ComputeSymbolMatchPriority(ISymbol symbol)
+        {
+            if (symbol.MatchesKind(SymbolKind.Local, SymbolKind.Parameter, SymbolKind.RangeVariable))
```

Comment:
And this ranking is a translation of the original implementation which was a function that compared 2 SymbolCompeltionItems to decide which to select.